plugin/plugin.go

```diff
@@ -363,6 +373,231 @@ func RegisterShutdown(name string, s Shutter) error {
+				return fmt.Errorf("in struct: %v", err)
+			}
+		}
+	case "slice":
+		if target.CanSet() != true {
+			return fmt.Errorf("unsettable value (unexported field?)")
+		}
+		// Get an Interface of the slice, extract a field and get the field's type
+		switch reflect.TypeOf(target.Interface()).Elem().Kind() {
+		case reflect.String:
+		case reflect.Float64:
+		case reflect.Bool:
+		default:
+			return fmt.Errorf("unsupported variable type (slices must be of string, float64, or bool)")
+		}
+	case "string":
+		if target.CanSet() != true {
+			return fmt.Errorf("unsettable value (unexported field?)")
+		}
+		return nil
+	case "float64":
+		if target.CanSet() != true {
+			return fmt.Errorf("unsettable value (unexported field?)")
+		}
+		return nil
+	case "bool":
+		if target.CanSet() != true {
+			return fmt.Errorf("unsettable value (unexported field?)")
+		}
+		return nil
+	default:
+		fmt.Println(target.Kind().String())
+		return fmt.Errorf("unsupported variable type (must be string, float64, bool, or slice of these, or struct)")
+	}
+	return nil // here to keep the compiler happy
+}
+
+// assignValue copies a oconfig_value to a target Value using the appropriate function
+func assignValue(oconfig_item *C.oconfig_item_t, target reflect.Value, index int) error {
+	valType, err := C.go_get_value_type(oconfig_item, C.int(index))
+	if err != nil {
+		return fmt.Errorf("unable to determine type of value")
+	}
+	switch valType {
+	case C.int(0): //string
+		target.SetString(C.GoString(C.go_get_string_value(oconfig_item, C.int(index))))
+	case C.int(1): // number/double/float64
+		target.SetFloat(float64(C.go_get_number_value(oconfig_item, C.int(index))))
+	case C.int(2): // boolean
+		target.SetBool(C.go_get_number_value(oconfig_item, C.int(index)) == 1)
+	default: // Unknown type
+		return fmt.Errorf("unsupported config item type")
+	}
+	return nil
+}
+
+// assignSlice generates a slice of the target type, loads in each value and assigns it to the target
+func assignSlice(oconfig_item *C.oconfig_item_t, target reflect.Value) error {
+	s := reflect.MakeSlice(target.Type(), int(oconfig_item.values_num), int(oconfig_item.values_num))
+	for i := 0; i < int(oconfig_item.values_num); i++ {
+		assignValue(oconfig_item, s.Index(i), i)
+	}
+	target.Set(s)
+	return nil
+}
+
+// assignConfig takes an oconfig_item_t struct and attempts to assign data to corresponding
+// fields in the provided config_target. A lack of corresponding field is not an error, a
+// corresponding field of the wrong type is.
+func assignConfig(oconfig *C.oconfig_item_t, target reflect.Value, isRoot bool) error {
+	// root oconfig_item's value is the name of the plugin and unused here
+	if isRoot != true {
+		if err := checkMatchable(oconfig, target); err != nil {
+			return fmt.Errorf("unmatchable value (%v)", err)
+		}
+	}
+	if int(oconfig.values_num) > 1 { // multiple values
+		assignSlice(oconfig, target)
+	} else if (oconfig.values_num == 1) && (isRoot == false) { // single value
```

Comment:
Good point regarding isRoot. 

You're right about <Option "with" "values">. I hadn't encountered/tested for that format which is a pity because as you say it breaks the model a little. So close and yet so far :)

I think the best struct is the one that can be instantly recognised by the developer as conforming with the required configuration. Is it fair to assume that plugins using
```
<Option "with" "values">
    Additionally "children"
</Option>
```
could in all cases be written as 
```
<Option "with">
    Additionally "children"
</Option>
<Option "values">
    Additionally "children"
</Option>
```
?

If so and I were writing a plugin I'd like to define:
```go
type ExamplePlugin struct { // maps to <Plugin "example"> by dint of being passed with that name
  Option map[string]struct { // a map of instances of this Option with the value as the key
    Additionally string
  }
}
```
This feels like it aligns better with the way that code utilising the configuration would be written.

If not or you feel it's too restrictive to expect configuration written that way I'd prefer to go back and do the work of using struct tags, if that would suit you. I think the rigidity of the current approach is useful for highlighting potential issues, but you know better what the community would prefer.